### ISSUE TITLE
@MockBean is not always resetting the mock after each test

### ISSUE THREAD

**Issue by GeorgiPetkov** [spring-projects/spring-boot#30625]
I've provided a minimal JUnit5 test `ConcreteServiceTest` reproducing the problem. Please check the code first. The issue is that the mock is not reset so the second test fails that there are 2 calls instead of the expected 1. The use case is having a common base service and similarly base test class.

Some observations:
- If the `@Import(ConcreteService.class)` is moved on top of the base class the tests pass.
- If the tests are flattened (not using `@Nested` test class) then the tests pass.
- `@AfterEach void clearMocks() { reset(dependency); }` would make the tests pass as well. I've chosen this as the simplest workaround for the time being (assuming the issue will be fixed and then the method will be simply deleted instead of altering the structure of my code).

Spring Boot version: 2.6.2

```
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.RepeatedTest;
import org.junit.jupiter.api.extension.ExtendWith;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.mock.mockito.MockBean;
import org.springframework.context.annotation.Import;
import org.springframework.test.context.junit.jupiter.SpringExtension;

import static org.mockito.Mockito.verify;

@Import(ConcreteService.class)
class ConcreteServiceTest extends BaseServiceTest {
}

@ExtendWith(SpringExtension.class)
abstract class BaseServiceTest {

    @MockBean
    private Runnable dependency;
    @Autowired
    private BaseService service;

    @Nested
    class Suite {

        @RepeatedTest(2)
        void test() {
            service.serve();

            verify(dependency).run();
        }
    }
}

class ConcreteService extends BaseService {
}

abstract class BaseService {

    @Autowired
    private Runnable dependency;

    public void serve() {
        dependency.run();
    }
}
```

**Comment by wilkinsona** [comment]
Thanks for the sample. The problem is occurring because, while the mock `Runnable` is being reset, it isn't the mock that's been injected into the `ConcreteServiceTest` and `ConcreteService` instances.

There are two application contexts involved here. The first is created for `ConcreteServiceTest` and the second is created for `Suite`. A mock is created and registered with each context. The mock from `ConcreteServiceTest`'s context is injected into the `dependency` field of both `ConcreteServiceTest` and `ConcreteService` but it's the mock from `Suite`'s context that is reset as this is the context that's "active" when executing `test()` on `Suite`.

I'm not sure what we can do about this in Boot. @sbrannen, can you please advise?

**Comment by sbrannen** [comment]
> I'm not sure what we can do about this in Boot. @sbrannen, can you please advise?

There's effectively nothing (reasonable) that we can do about it in either Spring Boot or Spring Framework.

The issue is that the `Suite` class inherits its configuration from its **enclosing class** which is `BaseServiceTest`, and `BaseServiceTest` is not annotated with `@Import(ConcreteService.class)`. Thus, the `MergedContextConfiguration` is different for `Suite` and `ConcreteServiceTest`. This causes those two test classes to have a different `ApplicationContext`.

The `MergedContextConfiguration` (which is used as the context cache key in the _Spring TestContext Framework_ (TCF)) is built based on static analysis of the test classes, their super classes, and their enclosing classes. Consequently, there is a disconnect between the _enclosing class_ for `Suite` and the concrete type of the _enclosing instance_ at runtime. At runtime, the enclosing instance is of type `ConcreteServiceTest`, but the enclosing class is of type `BaseServiceTest`.

With the current infrastructure in the TCF, all lookups for the enclosing type are based on the enclosing class (retrieved via `java.lang.Class.getEnclosingClass()`). In theory it should be possible to use `org.junit.jupiter.api.extension.ExtensionContext.getTestInstances()` to discover the concrete type of the enclosing instance; however, that would require an exhaustive rewrite of the internals in the TCF, and even if that were done... there would likely be other parts of the Spring portfolio or Spring community where enclosing type lookups are hard coded to use the enclosing class.

Moreover, as far as I know, the standard Java APIs do not support such lookups (for example, discovering the concrete type of an enclosing instance via reflection).

In summary, I would advise against `@Nested` test class structures where Spring annotations are only present on a subclass of a class whose nested test classes depend on the annotations of that concrete subclass.

----

For what it's worth, this scenario is not limited to the use of `@Import`. The same scenario arises if you replace `@Import(ConcreteService.class)` with `@ContextConfiguration(classes = ConcreteService.class)` (or similar arrangements of test-related annotations that affect the context cache key).

**Comment by wilkinsona** [comment]
Thanks very much for talking a look, @sbrannen.

Closing based on the recommendations above.

**Comment by GeorgiPetkov** [comment]
@wilkinsona I'm just wondering if this is the correct handling of the issue. It's obviously a real issue. A recommendation here in a closed issue is almost worthless. Even if this is not fixed any time soon this could be useful in the case of later redesign.

I believe there are still other things that can be done like documenting this somewhere or even better - detecting this case so there's a proper error/warning message for this. I've spent hours trying to figure out what is wrong because the usage was actually correct, but the implementation is unintendedly not supporting it and there is no indication of that whatsoever.

I understand that this is not a top priority issue, but just closing it seems inappropriate to me.

**Comment by wilkinsona** [comment]
It may be worth documenting this advice from Sam:

>  I would advise against `@Nested` test class structures where Spring annotations are only present on a subclass of a class whose nested test classes depend on the annotations of that concrete subclass.

It way even be possible to detect this arrangement and log a warning or throw an exception. It's something that would be best done in Spring Framework as it's general advice for using the Test Context Framework.

We'll transfer this issue to the Spring Framework issue tracker so that they can consider some documentation improvements.
